consensus/istanbul/backend/handler.go

```diff
@@ -98,28 +142,74 @@ func (sb *Backend) SetBroadcaster(broadcaster consensus.Broadcaster) {
 	sb.broadcaster = broadcaster
 }
 
-func (sb *Backend) NewChainHead() error {
+// SetP2PServer implements consensus.Handler.SetP2PServer
+func (sb *Backend) SetP2PServer(p2pserver consensus.P2PServer) {
+	sb.p2pserver = p2pserver
+	sb.valEnodeTable.p2pserver = p2pserver
+}
+
+// This function is called by worker.go whenever it gets a newWork event.
+func (sb *Backend) NewWork() error {
 	sb.coreMu.RLock()
 	defer sb.coreMu.RUnlock()
 	if !sb.coreStarted {
 		return istanbul.ErrStoppedEngine
 	}
 
-	// If the last block of the epoch has just been added to the blockchain, then
-	// establish 'validator' type connections to all validators in the upcoming epoch
-	// and disconnect from the ones that are no longer in the val set.
-	currentBlock := sb.currentBlock()
-	if istanbul.IsLastBlockOfEpoch(currentBlock.Number().Uint64(), sb.config.Epoch) {
-		sb.logger.Trace("At end of epoch and going to refresh validator peers", "current block number", currentBlock.Number().Uint64())
-		valset := sb.getValidators(currentBlock.Number().Uint64(), currentBlock.Hash())
-		if _, val := valset.GetByAddress(sb.Address()); val == nil {
+	go sb.istanbulEventMux.Post(istanbul.FinalCommittedEvent{})
```

Comment:
Previously, the val_enode_table (and the validator peers) were refreshed on epoch start ONLY on nodes that have `--mine` specified since `NewChainHead` was called within `worker.go`.

Now, sentries will need to do that refresh, and they don't have `--mine` turned on.

Reply:
Sorry, should of been more clear.  The function that originally was named `NewChainHead` is now renamed to `NewWork`.  `NewWork` will be called by only nodes with `--mine` enabled.

The "new" `NewChainHead` function will be called by called by full nodes and miners.